### PR TITLE
Add generation script for LoRA

### PR DESCRIPTION
**Pull request by lantiga**
Doesn't support `quantize` yet.

### CODE REVIEW

File: generate_lora.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    temperature: float = 0.8,
+    accelerator: str = "auto",
+) -> None:
+    """Generates a response based on a given instruction and an optional input.
+    This script will only work with checkpoints from the instruction-tuned LoRA model.
+    See `finetune_lora.py`.
+
+    Args:
+        prompt: The prompt/instruction (Alpaca style).
+        lota_path: Path to the checkpoint with trained LoRA weights, which are the output of
```
Comment (awaelchli) [inline generate_lora.py:41]:
```suggestion
        lora_path: Path to the checkpoint with trained LoRA weights, which are the output of
```

File: generate_lora.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import warnings
+from pathlib import Path
+from typing import Optional
+
+import lightning as L
+import torch
+
+from generate import generate
+from lit_llama import Tokenizer
+from lit_llama.adapter import LLaMA, LLaMAConfig
```
Comment (awaelchli) [inline generate_lora.py:12]: The import needs to be fixed

File: generate_lora.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        dtype: The dtype to use during generation.
+        max_new_tokens: The number of generation steps to take.
+        top_k: The number of top most probable tokens to consider in the sampling process.
+        temperature: A value controlling the randomness of the sampling process. Higher values result in more random
+            samples.
+        accelerator: The hardware to run on. Possible choices are:
+            ``"cpu"``, ``"cuda"``, ``"mps"``, ``"gpu"``, ``"tpu"``, ``"auto"``.
+    """
+    if not lora_path:
+        lora_path = Path("out/adapter/alpaca/lit-llama-lora-finetuned.pth")
```
Comment (awaelchli) [inline generate_lora.py:58]:
```suggestion
        lora_path = Path("out/lora/alpaca/lit-llama-lora-finetuned.pth")
```
And the finetune_lora is a bit different from finetune-alpaca as we don't write an end-of-trianing checkpoint

File: generate_lora.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        pretrained_path = Path(f"./checkpoints/lit-llama/7B/lit-llama.pth")
+    if not tokenizer_path:
+        tokenizer_path = Path("./checkpoints/lit-llama/tokenizer.model")
+    
+    assert lora_path.is_file()
+    assert pretrained_path.is_file()
+    assert tokenizer_path.is_file()
+
+    if quantize is not None:
+        raise NotImplementedError("Quantization in LoRA is not supported yet")
```
Comment (awaelchli) [inline generate_lora.py:69]: Should we remove the argument then? In adapter it is also not supported atm.

File: generate_lora.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    ), lora(r=lora_r, alpha=lora_alpha, dropout=lora_dropout, enabled=True):
+        print("Loading model ...", file=sys.stderr)
+        t0 = time.time()
+        model = LLaMA(LLaMAConfig())  # TODO: Support different model sizes
+
+        # 1. Load the pretrained weights
+        pretrained_checkpoint = torch.load(pretrained_path)
+        model.load_state_dict(pretrained_checkpoint, strict=False)
+
+        # 2. Load the fine-tuned adapter weights
```
Comment (awaelchli) [inline generate_lora.py:89]:
```suggestion
        # 2. Load the fine-tuned LoRA weights
```

**Review by awaelchli — APPROVED** [review]

File: generate_lora.py (hunk 3 above)
Comment (lantiga) [reply generate_lora.py:58]: I did add saving in finetune_alpaca.py

File: generate_lora.py (hunk 4 above)
Comment (lantiga) [reply generate_lora.py:69]: I kept it since it's probably the next thing we need to look into

**Pull request merged**